devices.go

```diff
@@ -146,29 +153,61 @@ func (d *Device) GetPortByName(name string) (*Port, error) {
 	return nil, fmt.Errorf("Port %s not found in device %s", name, d.ID)
 }
 
-func (d *Device) GetNetworkType() (string, error) {
-	numOfBonds := d.NumOfBonds()
-	if (numOfBonds < 1) || (numOfBonds > 2) {
-		return "", fmt.Errorf("Wrong number of Bond ports")
+type ports map[string]*Port
+
+func (ports ports) allBonded() bool {
+	if ports == nil {
+		return false
 	}
-	bond0, err := d.GetPortByName("bond0")
-	if err != nil {
-		return "", err
+
+	if len(ports) == 0 {
+		return false
 	}
-	if numOfBonds == 2 {
-		bond1, err := d.GetPortByName("bond1")
-		if err != nil {
-			return "", err
+
+	for _, p := range ports {
+		if (p == nil) || (!p.Data.Bonded) {
+			return false
 		}
-		if bond0.NetworkType == bond1.NetworkType {
-			return bond0.NetworkType, nil
+	}
+	return true
+}
+
+func (d *Device) HasManagementIPs() bool {
+	for _, ip := range d.Network {
+		if ip.Management {
+			return true
+		}
+	}
+	return false
+}
+
+// GetNetworkType returns a composite network type identification for a device
+// based on the plan, network_type, and IP management state of the device.
+// GetNetworkType provides the same composite state rendered in the Packet
+// Portal for a given device.
+func (d *Device) GetNetworkType() string {
+	if d.Plan != nil {
+		if d.Plan.Slug == "baremetal_0" || d.Plan.Slug == "baremetal_1" {
+			return NetworkTypeL3
 		}
-		if (bond0.NetworkType == "layer3") && (bond1.NetworkType == "layer2-individual") {
-			return "hybrid", nil
+		if d.Plan.Slug == "baremetal_1e" {
+			return NetworkTypeHybrid
+		}
+	}
+
+	bonds := ports(d.GetBondPorts())
+	phys := ports(d.GetPhysicalPorts())
+
+	if bonds.allBonded() {
+		if phys.allBonded() {
+			if !d.HasManagementIPs() {
+				return NetworkTypeL2Bonded
+			}
+			return NetworkTypeL3
 		}
-		return "", fmt.Errorf("Strange 2-bond ports conf - bond0: %s, bond1: %s", bond0.NetworkType, bond1.NetworkType)
```

Comment:
@t0mk On second thought, I now realize that hardware can sometimes be misconfigured (installed) or some NICs may not physically activate. I'm considering adding another network state "Invalid-ContactSupport" for these very special cases (for example, eth2 and eth3 were present for one user while eth0 and eth1 were not).  These checks would be based on some early assertions.  The network_type request would not return an error, just this state.

Convert functions would return an error, because they can not be expected to perform in these conditions. 

I'm not sure if an "Unknown" state may also be needed. I think the logic that we get from the Web UI is pretty solid in that regard.  Even the Web UI was confused when this physically misconfigured device presented without eth0/eth1.

Reply:
Yes, I remember the Web UI didn't handle certain state combination of the eths. The fallback wasn't very informative.

The "contact support" state sounds like a good idea, but then the error handling and propagation would have to be rethought, especially in terms of the network state/type transitions. I can't dive deep to it right now, but what you've wrote sounds good.